torch/lib/THCS/generic/THCSTensorMath.cu

```diff
@@ -217,21 +217,25 @@ void THCSTensor_(hspmm)(THCState *state, THCSTensor *r_, real alpha, THCSTensor
 void THCSTensor_(spcadd)(THCState *state, THCTensor *r_, THCTensor *dense, real value, THCSTensor *sparse) {
   THCAssertSameGPU(THCSTensor_(checkGPU)(state, 1, 3, sparse, r_, dense));
 
+  const ptrdiff_t nnz = THCSTensor_(nnz)(state, sparse);
+  if (nnz == 0) {
+    THCTensor_(resizeAs)(state, r_, dense);
+    THCTensor_(copy)(state, r_, dense);
```

Comment:
I suggest:

{{{
diff --git a/torch/lib/TH/generic/THTensorCopy.c b/torch/lib/TH/generic/THTensorCopy.c
index 71ccfdd..6e08ba6 100644
--- a/torch/lib/TH/generic/THTensorCopy.c
+++ b/torch/lib/TH/generic/THTensorCopy.c
@@ -68,7 +68,9 @@ void THTensor_(copyTranspose)(THTensor *tensor, THTensor *src) {
 
 void THTensor_(copy)(THTensor *tensor, THTensor *src)
 {
-  if (THTensor_(isContiguous)(tensor) && THTensor_(isContiguous)(src) && THTensor_(nElement)(tensor) == THTensor_(nElement)(src)) {
+  if (tensor == src) {
+    // Nothing to do
+  } else if (THTensor_(isContiguous)(tensor) && THTensor_(isContiguous)(src) && THTensor_(nElement)(tensor) == THTensor_(nElement)(src)) {
     real *sp = THTensor_(data)(src);
     real *rp = THTensor_(data)(tensor);
     ptrdiff_t sz = THTensor_(nElement)(tensor);
diff --git a/torch/lib/THC/generic/THCTensorCopy.cu b/torch/lib/THC/generic/THCTensorCopy.cu
index 4198025..f99ea44 100644
--- a/torch/lib/THC/generic/THCTensorCopy.cu
+++ b/torch/lib/THC/generic/THCTensorCopy.cu
@@ -4,7 +4,9 @@
 
 THC_API void
 THCTensor_(copy)(THCState* state, THCTensor* dst, THCTensor* src) {
-  THC_copyTensor<THCTensor, THCTensor>(state, dst, src);
+  if (dst != src) {
+    THC_copyTensor<THCTensor, THCTensor>(state, dst, src);
+  }
 }
 
 THC_API void
}}}